tests/test_resume2.rs

```diff
@@ -343,11 +343,7 @@ impl DataSource<u64> for TestSource {
 
 struct InsertDataSyscall(Arc<Mutex<Snapshot2Context<u64, TestSource>>>);
 
-impl<Mac: SupportMachine> Syscalls<Mac> for InsertDataSyscall {
-    fn initialize(&mut self, _machine: &mut Mac) -> Result<(), Error> {
-        Ok(())
-    }
-
+impl<Mac: SupportMachine> ExecutionContext<Mac> for InsertDataSyscall {
```

Comment:
I wasn't sure on this, perhaps @mohanson's rejection partially comes from this: I could get the benefits of having `ExecutionContext` handling lifetimes, but I do believe the merits of keeping the old Syscalls trait for more complicated syscalls. Right now this PR would push for the following cases:

* Simpler syscalls must be changed to pure functions
* More complicated syscalls must implement ExecutionContext instead

What if we keep a middle ground where thread-safe syscalls can safely use Syscalls trait? One can also directly implement ExecutionContext for the referenced types, but having Syscalls in the codebase definitely aids backwards compatibility. Even better, we can keep a thread safe version of Syscalls trait and a non-thread safe version of NonThreadSafeSyscalls trait.

Simply removing Syscalls trait, IMHO, is not a good choice judging from both compatibility and code clarity reasons. Syscalls are syscalls, developers should only deal with syscalls when they needed, maybe ExecutionContext can mostly remain as a transparent concept?

Reply:
Syscall impls don't need to be changed to pure functions. Just implement `ExecutionContext` instead of `Syscall`.

I agree that we can keep the `Syscall` trait for better compatibility, for stacking multiple `Syscall` impls like ckb.

But it will have nothing to do with thread safety. It shouldn't be necessary to declare `Syscall: Sync + Sync`. The machine types are `Send`/`Sync` when the type params are.

Reply:
Let's break it down:

* I don't think syscalls should implement `ExecutionContext`, syscalls should simply implement Syscalls trait. In fact, I want to hide `ExecutionContext` as much as possible. Users of ckb-vm should not need to know `ExecutionContext` in 99% of the cases, all they need to deal with, is methods in DefaultMachineBuilder. One example here, is I think we should see `ExecutionContext` much like the parameter `A` used to declare allocator in [Vec](https://doc.rust-lang.org/std/vec/struct.Vec.html)
* The fundamental goal here, is we want to make sure `AsmMachine` as well as `TraceMachine` are thread safe, so we can safely move them among threads in CKB. It is just due to the previous design that Syscalls must be declared as `Sync + Send` to achieve this goal. So as long ias `AsmMachine` / `TraceMachine` can be `Send + Sync` with the help of `ExecutionContext`, I think we can safely remove `Sync + Send` on Syscalls trait directly, but I wasn't so familiar with lifetime gymnastics to draw a conclusion here. That's why I suggested a pair of Syscalls / NonThreadSafeSyscalls in previous comment. If you think AsmMachine can be `Send + Sync` even if Syscalls traits are not `Send + Sync`, I'm personally good with removing those type parameters on Syscalls.

Reply:
* It's not like really `Allocator`, where the default global one is used 99% of the times. When you don't need to name to machine type, it doesn't matter, you can just use the builder methods to compose a machine. When you do, you need to decide whether you want to define a type that implements `ExecutionContext` or use `Box<dyn>` or maybe use impl trait if it's just in return type.
* It's not necessary. All the tests that move a machine to another thread are still working. And if you generate docs for this you'll see:

impl<Inner, ExecutionContext> [Send](https://doc.rust-lang.org/1.67.1/core/marker/trait.Send.html) for [DefaultMachine](file:///home/ygh/src/ckb-vm/target/doc/ckb_vm/machine/struct.DefaultMachine.html)<Inner, ExecutionContext>where
    ExecutionContext: [Send](https://doc.rust-lang.org/1.67.1/core/marker/trait.Send.html),
    Inner: [Send](https://doc.rust-lang.org/1.67.1/core/marker/trait.Send.html),
impl<Inner, ExecutionContext> [Sync](https://doc.rust-lang.org/1.67.1/core/marker/trait.Sync.html) for [DefaultMachine](file:///home/ygh/src/ckb-vm/target/doc/ckb_vm/machine/struct.DefaultMachine.html)<Inner, ExecutionContext>where
    ExecutionContext: [Sync](https://doc.rust-lang.org/1.67.1/core/marker/trait.Sync.html),
    Inner: [Sync](https://doc.rust-lang.org/1.67.1/core/marker/trait.Sync.html),

Reply:
Even when using `Box<dyn>` it shouldn't be necessary to declare `Syscalls: Send + Sync`. Instead use `Box<dyn Syscalls + Send + Sync>`.

Reply:
I have restored the `Syscalls` trait and `impl Syscalls for Box<T> where T: Syscalls`. So existing code that uses only `syscall` and `instruction_cycle_func` (this is because `Box<F>` implements `Fn(...)` if `F` implements `Fn(...)`) and doesn't care about the machine type will work without modification, like the vm runner example and many tests. (But it'll be static dispatching instead of dynamic dispatching. It's also recommended to remove the `Box` if it isn't really necessary.)

Reply:
> It's not like really Allocator, where the default global one is used 99% of the times. When you don't need to name to machine type, it doesn't matter, you can just use the builder methods to compose a machine. When you do, you need to decide whether you want to define a type that implements ExecutionContext or use Box<dyn> or maybe use impl trait if it's just in return type.

We will have to agree to disagree, the above arguments, to me, is proof that ExecutionContext still is leaky abstraction that requires slightly more work. Yes in the new builder design, from now and then there might be the need to explicitly specify a type, but my question is: does one need to specify a full `ExecutionContext` type? Can we stop at syscall or debugger level?

> All the tests that move a machine to another thread are still working.

I wasn't sure on this, do we have tests like this in this repository?

Reply:
The alternative is to have four type parameters, which is IMO a bit too many.

Or maybe keep using `Box<dyn>` for debugger/cycles function like the previous PR. But it's actually really limiting because you are then restricted to `Send + Sync + 'static` Debuggers. IMO it's more leaky than having a single context trait.

We do have tests with threads, just grep for spawn.

Reply:
I don't really see why we need 4 type parameters.

Reply:
I mean `DefaultMachine<Inner, Syscalls, Debugger, InstCyclesFunc>`.